R/util_dfc.R

```diff
@@ -0,0 +1,238 @@
+  # Otherwise, duplicate ID values within each data frame will get ignored.
+  # Example use case: you want to compare a student-x-week df with a student df to see if they have the same students.
+  # The commonly-shared id_col is student ID.
+  # Since student ID doesn't uniquely identify rows in the first df, id_cols_uniquely_identify_rows = FALSE.
+
+  # Sanity checks
+  if(!is.data.frame(df1) | !is.data.frame(df2)) {
+    stop("Error - at least one of the first two arguments for util_dfc.compare_unique_identifiers is not a data frame.")
+  }
+  if(is.na(id_cols) | length(id_cols) == 0) {
+    stop("Error - id_cols argument is NA or empty.")
+  }
+  if(any(!id_cols %in% names(df1)) | any(!id_cols %in% names(df2))) {
+    stop("Error - not all id columns provided are in both data frames.")
+  }
+
+  # Extract the IDs from df1 and df2 in character format.
+  # If IDs are defined by multiple columns, use them all with "__" as a separator.
+  # Note: this section treats NAs as the character string "NA" in an ID.
+  df1_ids <- util_dfc.get_condensed_ids(df1, id_cols)
+  df2_ids <- util_dfc.get_condensed_ids(df2, id_cols)
+
+  # If the IDs aren't expected to uniquely identify rows, then cut duplicates from them.
+  if(!id_cols_uniquely_identify_rows) {
+    df1_ids <- df1_ids[!duplicated(df1_ids)]
+    df2_ids <- df2_ids[!duplicated(df2_ids)]
+  }
+
+  # Warn if the IDs still have duplicates in either df!
+  # This would mean that the user expected no duplicates but actually had them.
+  if(any(duplicated(df1_ids))) {
+    warning("Warning - first data frame passed to util_dfc.compare_unique_identifiers has unexpected duplicate ID values:")
+    util.print_pre(unique(df1_ids[duplicated(df1_ids)]))
+  }
+  if(any(duplicated(df2_ids))) {
+    warning("Warning - second data frame passed to util_dfc.compare_unique_identifiers has unexpected duplicate ID values:")
+    util.print_pre(unique(df2_ids[duplicated(df2_ids)]))
+  }
+
+  # use util_dfc.compare_vecs on id_cols of each df
+  return(util_dfc.compare_vecs(df1_ids, df2_ids))
+}
+
+
+
+
+util_dfc.compare_dfs <- function(df1, df2, id_cols = c()) {
+  # Big wrapper function for giving an overall picture of how two DFs differ:
+  ## Check nrow and ncol
+  ## Compare colnames
+  ## Compare identifiers
+  ## Subset to shared columns, sort columns and rows (using id_cols), and report which values match exactly in the subset.
+
+  ##### BASIC SANITY-CHECKS AND COMPARISONS
+
+  # Setup
+  df1_name <- deparse(substitute(df1))
+  df2_name <- deparse(substitute(df2))
+
+  # Are they data frames?
+  if(!is.data.frame(df1) | !is.data.frame(df2)) {
+    stop("Error - at least one of the first two arguments for util_dfc.compare_dfs is not a data frame.")
+  }
+
+  # Compare dimensions
+  util.print_pre("COMPARING DIMENSIONS OF DATA FRAMES: ")
+  util.print_pre("Data frame 1: " %+% paste(dim(df1), collapse = " x "))
+  util.print_pre("Data frame 2: " %+% paste(dim(df2), collapse = " x "))
+
```

Comment:
If we use these functions in non-interactive mode, it could be annoying not to be able to turn off util.print_pre. It might make more sense here to just use `message()` or something that can be turned off in rmarkdown chunks if desired.

Reply:
Cool - switching to message().